features/steps/steps.py

```diff
@@ -64,36 +64,116 @@ def given_a_machine(context, series):
         CONTAINER_PREFIX + pr_prefix + vm_prefix + series + date_prefix
     )
 
-    context.instance = context.config.cloud_manager.launch(
-        series=series,
-        instance_name=instance_name,
-        image_name=context.series_image_name[series],
-    )
+    context.instances = {
+        "uaclient": context.config.cloud_manager.launch(
+            series=series,
+            instance_name=instance_name,
+            image_name=context.series_image_name[series],
+        )
+    }
 
     context.container_name = context.config.cloud_manager.get_instance_id(
-        context.instance
+        context.instances["uaclient"]
     )
 
     def cleanup_instance() -> None:
         if not context.config.destroy_instances:
             print(
                 "--- Leaving instance running: {}".format(
-                    context.instance.name
+                    context.instances["uaclient"].name
                 )
             )
         else:
             try:
-                context.instance.delete(wait=False)
+                context.instances["uaclient"].delete(wait=False)
             except RuntimeError as e:
                 print(
                     "Failed to delete instance: {}\n{}".format(
-                        context.instance.name, str(e)
+                        context.instances["uaclient"].name, str(e)
                     )
                 )
 
     context.add_cleanup(cleanup_instance)
-    logging.warning("--- instance ip: {}".format(context.instance.ip))
-    print("--- instance ip: {}".format(context.instance.ip))
+    logging.warning(
+        "--- instance ip: {}".format(context.instances["uaclient"].ip)
+    )
+    print("--- instance ip: {}".format(context.instances["uaclient"].ip))
+
+
+@when("I launch a `{series}` `{instance_name}` machine")
```

Comment:
This is solid simple and flexible, at some point it feels like we could extend this to
`@when("I launch a `{series}` `{instance_name}` machine with {userdata}")`
And then we can just provide the #cloud-config needed in the `context.text` to feed into the VM to cut down on all the natural language for setting up files etc.

Reply:
Completely agree, we should allow that flexibility in the future